services/drupal/web/modules/custom/epa_core/epa_core.deploy.php

```diff
@@ -433,18 +433,157 @@ function epa_core_deploy_0004_set_card_field_default_values(&$sandbox) {
           \Drupal::logger('epa_core')->error("Error inserting record into $field_table table: " . $e->getMessage());
         }
       }
-    else {
-      // Record does exist but is null, run update.
-      try {
-        $database->update($field_table)->fields([
-          'field_flag_card_alignment_value' => 'default'
-        ])->execute();
-      }
-      catch (\Exception $e) {
-        \Drupal::logger('epa_core')->error("Error updating record into $field_table table: " . $e->getMessage());
+      else {
+        // Record does exist but is null, run update.
+        try {
+          $database->update($field_table)->fields([
+            'field_flag_card_alignment_value' => 'default'
+          ])->execute();
+        }
+        catch (\Exception $e) {
+          \Drupal::logger('epa_core')->error("Error updating record into $field_table table: " . $e->getMessage());
+        }
+
       }
+    }
+  }
+}
+
+/**
+ * Setting notification_opt_in flag for node author, author of the latest
+ * revision, or author of the current revision
+ */
+function epa_core_deploy_0005_set_watch_flag(&$sandbox) {
+  if (!isset($sandbox['total'])) {
+    // Get all nodes.
+    $nodes = \Drupal::database()->query(
+      "SELECT node.nid,
+         latest.revision_uid,
+         node.uid,
+         current.revision_uid
+FROM node_revision AS latest
+LEFT JOIN node_field_data AS node
+    ON latest.nid = node.nid
+LEFT JOIN node_revision AS current
+    ON node.vid = current.vid
+WHERE latest.vid IN
+    (SELECT MAX(vid)
+    FROM node_revision
+    GROUP BY  nid);")
+      ->fetchAll();
+
+    $sandbox['total'] = count($nodes);
+    $sandbox['current'] = 0;
+    \Drupal::logger('epa_core')
+      ->notice('Queueing ' . count($nodes) . ' nodes to be updated with notification_opt_in flag');
+  }
+
+  // Query 500 at a time for batch.
+  $nodes = \Drupal::database()->query(
+    "SELECT node.nid,
+         latest.revision_uid as latest_uid,
+         node.uid,
+         current.revision_uid
+FROM node_revision AS latest
+LEFT JOIN node_field_data AS node
+    ON latest.nid = node.nid
+LEFT JOIN node_revision AS current
+    ON node.vid = current.vid
+WHERE latest.vid IN
+    (SELECT MAX(vid)
+    FROM node_revision
+    GROUP BY  nid) LIMIT 500
+        OFFSET " . $sandbox['current'] . ";")
+    ->fetchAll();
+
+  foreach ($nodes as $data) {
+    _epa_core_set_notification_opt_in_flag($data);
+    $sandbox['current']++;
```

Comment:
Rather than tracking the number you've processed it would be better to track the high water mark of the last value of the nid you've processed. This is superior because if nodes are deleted while you're in the process of running these batches a simple "count" could end up skipping records. To explain consider a simple example where you're grabbing 3 nodes at a time from this list of IDs:
- 1
- 2
- 3
- 4
- 5
- 6
- 7

Your first batch processes 1, 2, and 3.  If ID 2 is deleted before your code grabs the 2nd batch, then the list looks like:
- 1
- 3
- 4
- 5
- 6
- 7

 second batch will grab nodes 5,6,7 because your offset of 3 skips the first 3 in the list.  In this scenario you end up skipping ID 4 altogether.

If you, instead, track the highest nid you've processed (ID 3 in this example) then your second request will be for the next 3 IDs that are higher than 3 and it'll correctly return records 4, 5, and 6.

Reply:
@drfuzetto I'd still make the recommendation that you switch to using a high-water-mark approach to manage pagination and for assessing completion rather than just a simple count-based approach.